# HG changeset patch
# User welpy-cw <h.w.forms@arcor.de>
# Date 1722960850 -3600
# Node ID 67d373e139953ed7949aed73ee6943f05db9d5df
# Parent  2c74b53a08ddb6cab0d923ea09a9b7f75a7d11ba
Bug 1909617 - Don't show "Mark Folder Read" for virtual folders with search terms. r=aleca

Without a view, there seems to be no reasonable way to mark as read only the searched messages in
the underlying folders. So I suggest this approach. The functionality is still provided by the
"Mark All Read" command when the folder is displayed (which should be fixed in bug 1909617).

Differential Revision: https://phabricator.services.mozilla.com/D218486

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -572,16 +572,25 @@ var folderPaneContextMenu = {
       folder;
     const isJunk = flags & Ci.nsMsgFolderFlags.Junk;
     const isTrash = isSpecialFolder(Ci.nsMsgFolderFlags.Trash, true);
     const isVirtual = flags & Ci.nsMsgFolderFlags.Virtual;
     const isRealFolder = !isServer && !isVirtual;
     const isSmartVirtualFolder = FolderUtils.isSmartVirtualFolder(folder);
     const isSmartTagsFolder = FolderUtils.isSmartTagsFolder(folder);
     const serverType = server.type;
+    const hasNoSearchTerms = () => {
+      if (!isVirtual) {
+        return true;
+      }
+      const wrapper = VirtualFolderHelper.wrapVirtualFolder(folder);
+      const noSearchTerms = ["", "ALL"].includes(wrapper.searchString);
+      wrapper.cleanUpMessageDatabase();
+      return noSearchTerms;
+    };
 
     this._showMenuItem(
       "folderPaneContext-getMessages",
       (isServer && serverType != "none") ||
         (["nntp", "rss"].includes(serverType) && !isTrash && !isVirtual)
     );
     const showPauseAll = isServer && FeedUtils.isFeedFolder(folder);
     this._showMenuItem("folderPaneContext-pauseAllUpdates", showPauseAll);
@@ -631,17 +640,20 @@ var folderPaneContextMenu = {
               ? "newFolder"
               : "newSubfolder"
           )
         );
     }
 
     this._showMenuItem(
       "folderPaneContext-markMailFolderAllRead",
-      !isServer && !isSmartTagsFolder && serverType != "nntp"
+      !isServer &&
+        !isSmartTagsFolder &&
+        hasNoSearchTerms() &&
+        serverType != "nntp"
     );
     this._showMenuItem(
       "folderPaneContext-markNewsgroupAllRead",
       isRealFolder && serverType == "nntp"
     );
     this._showMenuItem(
       "folderPaneContext-emptyTrash",
       isSpecialFolder(Ci.nsMsgFolderFlags.Trash, true)
